### PR TITLE
Add delete tag endpoint

### PR DESCRIPTION
**Pull request by nojaf**
I've add a simple endpoint and some steps of how I did it.
Please review and don't be shy to add remarks on the markdown!

### CODE REVIEW

File: porting-a-new-endpoint.md
Comment (jescalada) [inline porting-a-new-endpoint.md:1]: A troubleshooting section might also be helpful. Any typical pitfalls you might encounter when following the steps should be documented to simplify new contributor's lives 🙂

**Comment by jescalada** [comment]
We could also add a section for performance-related considerations. Since we are porting to Go for performance reasons (?), there are times where we don't want to do things exactly like they're done in Python? Maybe we can add some actual examples of rewriting existing logic to make it more efficient or idiomatic.

File: porting-a-new-endpoint.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+type TrackingService interface {
+    ...
+	DeleteTag(ctx context.Context, input *protos.DeleteTag) (*protos.DeleteTag_Response, *contract.Error)
+}
+```
+
+### New FFI Endpoint
+
+In [pkg/lib/tracking.g.go](pkg/lib/tracking.g.go), we need to expose that same service method so it can be called directly from Python via FFI. More on that later.
+
```
Comment (suprjinx) [inline porting-a-new-endpoint.md:74]: The `.g.go` suffix indicates a generated file, right? why are we editing it manually?

File: porting-a-new-endpoint.md
Comment (suprjinx) [inline porting-a-new-endpoint.md:1]: Excellent documentation! Should it perhaps go in a `docs` folder?

File: porting-a-new-endpoint.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+```
+
+After that, run `mage generate` again and check if our fields in `DeleteTag` now contain `validate:"required"`.
+
+## Implement the Service Method
+
+We aim to keep the Go implementation as close to the Python code as possible. That's why we have a thin service layer and keep most logic in the (SQL) store instead of the service. This keeps things similar to Python, making it easier to port and compare the code.
+
+In the service, we need to call the store with the same arguments as the Python store would be called. This is typically where we convert proto structs to entities. In the case of `deleteTag`, we can just pass the two arguments to the store.
+
```
Comment (jescalada) [inline porting-a-new-endpoint.md:125]: It would be good to explain which file to edit, and whether we need to create a new file or not (in this case, we had to create the `/pkg/tracking/service/tags.go` file since there weren't any other tag endpoints created yet).

File: porting-a-new-endpoint.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+```go
+func (s TrackingSQLStore) DeleteTag(
+	ctx context.Context, runID string, key string,
+) *contract.Error {
+	return nil
+}
+```
+
+Lastly, update the `TrackingStore` interface mock via `go generate ./...`.
```
Comment (jescalada) [inline porting-a-new-endpoint.md:172]: I found this a bit confusing, as I thought the `...` part meant something like "Fill in the blank". Perhaps we could rephrase this and make it obvious that we want to run that specific command (by wrapping it in copy-pastable code snippet):

Lastly, run the following command to update the mock `TrackingStore` at `pkg/tracking/store/mock_tracking_store.go`:
```
go generate ./...
```

File: porting-a-new-endpoint.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+If everything goes well, you should be able to run `mage test:unit` to execute the Go unit tests. This serves as a sanity check to verify that the Go code is compiling.
+
+### The Actual Store Implementation
+
+At this point, we need to write the actual implementation. Again, we want to stay true to the Python code, so please refer to [sqlalchemy_store.py](.mlflow.repo/mlflow/store/tracking/sqlalchemy_store.py).
```
Comment (jescalada) [inline porting-a-new-endpoint.md:178]: Just a minor issue: clicking the link doesn't seem to take us to the file since it isn't actually part of the current repo. Can we perhaps link the file from an external repo, or simply show the file path?

File: porting-a-new-endpoint.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+It is important that the same data and exceptions are returned to ensure the Go implementation behaves like the Python one, allowing users to seamlessly switch to Go while having the same experience.
+
+Don’t hesitate to challenge the current implementation of the Python code. You might find opportunities to revisit and improve parts of it.
+
```
Comment (jescalada) [inline porting-a-new-endpoint.md:183]: I found the process of porting the code a bit confusing, since the Python version looks succint compared to the Go translation. When running the tests, I noticed they often fail due to validation checks, and the `DeleteTag` implementation doesn't seem to explicitly handle validations as the Python code does.

When implementing the `SetTag`, I added explicit checks to try and get the tests to pass, but these are probably inefficient...

To summarize, a slightly deeper explanation on how to port the functions could be very useful 😃

File: porting-a-new-endpoint.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+
+## Write Tests
+
```
Comment (jescalada) [inline porting-a-new-endpoint.md:187]: An example of a unit test for a certain endpoint would be really valuable here to give contributors a template to work with

**Review by jescalada — COMMENTED** [review]
This is a second, in-depth review of the guide from going through all the steps as a Go newbie. Feel free to take them or leave them as needed :)

File: docs/porting-a-new-endpoint.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
+# Porting a New Endpoint
+
+In order to complete the Go implementation of the tracking server, we need to port the missing endpoints. 
+The following guide will walk you through the whole process of viewing the missing endpoints, implementing them and making sure they work with MLflow's existing integration tests.
+
+1. [View Missing Endpoints](#view-missing-endpoints): View the current state of implementated endpoints.
+2. [Enable the Endpoint](#enable-the-endpoint): Enable code generation for a new endpoint.
+3. [Validate Input](#validate-input): Apply validation on the incoming request data.
+4. [Implement the Service Method](#implement-the-service-method): Adding a new method to the TrackingService interface allows handling the logic for the new endpoint
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:9]:
```suggestion
4. [Implement the Service Method](#implement-the-service-method): Add a new method to the `TrackingService` interface to handle the logic for the new endpoint
```

File: docs/porting-a-new-endpoint.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        },
+    )
+    _get_tracking_store().delete_tag(request_message.run_id, request_message.key)
+    response_message = DeleteTag.Response()
+    response = Response(mimetype="application/json")
+    response.set_data(message_to_json(response_message))
+    return response
+```
+
+We notice that `run_id` and `key` are required. Too avoid writing repeatitive code in Go, we make use of [Go validator](https://github.com/go-playground/validator) and use annotations to validate the incoming structs.
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:112]:
```suggestion
We notice that `run_id` and `key` are required. To avoid writing repetitive code in Go, we make use of [Go validator](https://github.com/go-playground/validator) and use annotations to validate the incoming structs.
```

File: docs/porting-a-new-endpoint.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+var validations = map[string]string{
+    ...
+	"DeleteTag_RunId":                    "required",
+	"DeleteTag_Key":                      "required",
+}
+```
+
+After that, run `mage generate` again and check if our fields in `DeleteTag` now contain `validate:"required"`.
+
+Note that we do not need to process `_assert_string` because in Go all the structs are already statically typed.
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:126]:
```suggestion
Note that we do not need to process `_assert_string` because in Go, all structs are already statically typed.
```

File: docs/porting-a-new-endpoint.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+We aim to keep the Go implementation as close to the Python code as possible. That's why we have a thin service layer and keep most logic in the (SQL) store instead of the service. This keeps things similar to Python, making it easier to port and compare the code.
+
+In the [pkg/tracking/service/service.go](../pkg/tracking/service/service.go), we need to call the store with the same arguments as the Python store would be called:
+
+```python
+# handlers.py
+_get_tracking_store().delete_tag(request_message.run_id, request_message.key)
+```
+
+We split up the service over multiple files (one Go file per entity), sometimes it might not existing yet.
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:141]:
```suggestion
We split up the service over multiple files (one Go file per entity), sometimes it might not exist yet.
```

File: docs/porting-a-new-endpoint.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+There is one caveat to these tests; occasionally, they may have a Python bias, meaning that the tests pass due to Python's dynamic nature, while our Go tests might fail because they are strongly typed. Another issue may arise if the Python implementation does not consistently return the same error messages. Therefore, it may be necessary to submit a PR to [mlflow](https://github.com/mlflow/mlflow) to adjust the existing tests.
+
+Some examples include:
+- https://github.com/mlflow/mlflow/pull/13233
+- https://github.com/mlflow/mlflow/pull/13128
+- https://github.com/mlflow/mlflow/issues/12550
+
+## Troubleshooting
+
+If you encounter any difficulties, please feel free to open a draft PR and ask specific questions. Once your questions are answered, kindly update this section if you’ve learned something that should be included in this guide.
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:233]:
```suggestion
If you encounter any difficulties, please feel free to open a draft PR and ask specific questions. Once your questions are answered, kindly update this section if you’ve learned something that could be valuable for other contributors.
```

File: docs/porting-a-new-endpoint.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+}
+```
+
+After that, run `mage generate` again and check if our fields in `DeleteTag` now contain `validate:"required"`.
+
+Note that we do not need to process `_assert_string` because in Go all the structs are already statically typed.
+
+It may happen that the [baked-in validations](https://github.com/go-playground/validator?tab=readme-ov-file#baked-in-validations) of the Go validator are not sufficient. If this occurs, please create a [custom validation rule](https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Custom_Validation_Functions) in [pkg/validation/validation.go](../pkg/validation/validation.go).
+
+## Implement the Service Method
```
Comment (jescalada) [inline docs/porting-a-new-endpoint.md:130]:
```suggestion
## Implement the Service Logic
```

**Review by jescalada — APPROVED** [review]
Guide looks great! Just had some minor wording changes (and index fix).
